[Python] Implement new Concept Exercise: none #1031

mohanrajanr opened the pull request:
I have written this PR as per the contributing docs. 
Would like @DavidGerva, and someone from the python maintainers to review this PR.

Few points to add:

Although the exercise was just about None, I felt that the way to test the exercise would have been hard without having any reference to the Input.

1. If we expect to return None, a simple pass() statement on the function would also pass the test case.
2. Used a list to manipulate and understand the use of None.

I'm not sure the example for this exercise is appropriate and I am willing to do a full redo if required. Let me know. Thanks.

BethanyG commented on languages/python/exercises/concept/none/.docs/introduction.md:
````diff
@@ -0,0 +1,26 @@
+```
+
+This declares the variable as well as defines them. But when you want to just declare the variable and not define them at start you use a special built in called as `None`.
+
````
For the `introduction.md`, the goal is to focus only on the information needed to explain/define the concept and complete the exercise.  It should be as simple & direct as possible, with example usage.  So I think we can remove these paragraphs about variable declaration, and then go into more detail or nuance in the `after.md`.

mohanrajanr replied:
Noted. I will refer the docs from other Sources you have given me and alter the introduction.md. 

Although, I should reference them in the design.md ( as well as after.md ) right?

BethanyG replied:
**Correct!**  we definitely **do** want those additional resources available to the student in places like `hints.md` and `after.md`.  And we also want them in the `design.md`, so that we internally have  a record of what resources we used to build the exercise (_and what we might need to alter or update at a later time_), and what our thinking was around the concept/topic.

BethanyG commented on languages/python/exercises/concept/none/.docs/introduction.md:
````diff
@@ -0,0 +1,26 @@
+```
+
+In Python, `None` is a Singleton built-in which is used to declare a variable as well as can be used as `False`. Meaning if you want to check for a variable which is declared with `None`, It will be considered as `False`
+
````
I think it's good here to emphasize the "placeholder"/null nature of `None` -- that it represents an _**absence of value**_,  and also to mention that Python uses `None` when a return from a function is not specified.    But referring to variable declaration (_None can be a placeholder when you don't have a default value, but it really doesn't have a function beyond that_) confuses things here.  -- see [python docs](), [https://docs.python.org/3/library/constants.html#None], [David Beazly's Python Course](https://dabeaz-course.github.io/practical-python/Notes/02_Working_with_data/01_Datatypes.html) (_which mentions the conditional evaluation_), and  [educative.io None keyword](https://www.educative.io/edpresso/what-is-the-none-keyword-in-python) for some wording examples that are a little more direct.

 I also think it's dangerous to tell the student that `None` can be _**used as**_ `False` as a main case.  `None` _**evaluates**_ to `False` when used in a conditional check (_which is something we should cover in this exercise_), but `None` is not of type `bool`, and won't behave as such.  For cases where a `False` type is needed, `bool` type `False` should be used.  "Falsey" and "Truthy" are a different concept exercise.

BethanyG replied:
I also think we should save the discussion about `None` being a Singleton object for the `after.md` (_maybe link to the educative.io page, or some other good explanation??_).  It is very valuable information, but not really necessary for understanding basic usage or for completing this exercise.

mohanrajanr replied:
Got it. I will mode the discussion of None being a singleton object to after.md 

It is something very necessary to know but also a little tough to understand. I will refer some docs to bring it up more easier.

BethanyG replied:
Agree.  We might want to consider a follow-on concept exercise around `None` being a singleton, or its possible uses as a `sentinel value` -- but it feels like that might be under an OOP category, or some other more "advanced" scenario.  

But speaking to those uses in the `after.md` is a great start -- you can give examples, point the student to some articles or problems as "next steps", or link to projects or code where `None` is used in that fashion.  There are a lot of possibilities for what a student could explore after this basic run-through....we just don't want to chase them away with too much info at the very start.  😉

BethanyG commented:
Hi mohanrajanr,

👋  Welcome to Exercism.  😄   Thank you **very much** for all your work on this PR!  We really appreciate the help!   This is a solid start and with some work I think we can get this merged in pretty quickly.  YAY.  🎉 🌟 

 I've left some in-line comments, and will be leaving some more as I work through all the files - so my apologies that they might appear a bit out of order.  To reply to your two points above: 

> If we expect to return None, a simple pass() statement on the function would also pass the test case.
Used a list to manipulate and understand the use of None.

**Yes** - we could have a scenario where a student used `pass()`  instead of writing code - although we also need to trust that a student is here because they **want** to understand and learn the concepts.  If they "game" the system, that is their choice. Our "job" is to come up with a good exercise piece that lets students get familiar with all the core uses of `None` - and give them some support with tests they can use to solidify their understanding. 

My biggest struggle with `None` is how to give practical and useful examples and meaningful (_but short_) practice.  I really  like your idea of seat assignment/reservation.

🤔  But I do think we need to cover a few more scenarios than the two we have.  What if instead of a `list` to iterate through, the reservations were in a `dict` we supplied (_or the student constructs to our spec_), and the student has to write a function that checks if a list of tables or dates or reservations was "fully taken", "partially taken" or "not taken" (**e.g.**, if the reservation limit was 4, does the `dict` have 4, <4, or `None`)?  

We might be able to then set up a few more scenarios around calculating availability, or totals or helper functions where we could test for `None` or return `None` or default to `None`.  Just thinking out loud here - we might also be able to do the same thing with lists.   Thoughts?

mohanrajanr commented:
Hey @BethanyG, Thank you so much for your inputs.

> although we also need to trust that a student is here because they want to understand and learn the concepts

My bad, I did not think that way. We should trust the student that they will go with the right thing.

> the reservations were in a dict we supplied (or the student constructs to our spec).

Awesome thought. I will rewrite few more cases that can revolve around this.
1. Create a custom Variable and return it.
2. Pass an argument and counting the availability like you said.
3. "fully taken", "partially taken" or "not taken" - again like a counting from a dict of list {"room1": [], "room2": [1, None, 1]}. Good idea.
4. Return None from a function - or should we tell him to return a function object that when called will return none. ( I think this is too much of a stretch. I certainly this function objects would be out of scope ).

Let me know your thoughts. I also am thinking of taking a few more pull requests as a part of the fest. Although each one takes more time than a regular PR, I want to do justice to the Fest as well as Open source. If you have any PR's that needs doc changes, let me know, I can have that as last resort as well.

mohanrajanr commented:
Hey @BethanyG , I have pushed the changes after the updates you mentioned. Kindly let me know after you take a look. I also saw that there is a slack channel to discuss before opening PR's and other info. Is there a way to join that?. I think it would be easier to get approvals before making the changes.

Thanks!

BethanyG commented:
@mohanrajanr,

Apologies -- getting a late start this Sunday (I am in Pacific Time).  I will take a look at your changes, and get back to you with feedback by EOD today my time.  Thank you again for all your work on this -- I also saw from your comments on issues that you also wanted to take on `loops` -- that is **awesome** . 😄  

It is really nice to see someone setting a goal of _actually helping_ during Hacktoberfest (_instead of submitting spam_).  You are making us very happy campers.

**Re:**  Slack -- I sent an invite to the email listed on your GitHub profile.  Let me know if you'd like me to send it elsewhere.   My user in Slack is the same as it is here - so drop in, introduce yourself - and we can have discussions in #maintaining-python and #track-python!

BethanyG commented on languages/python/exercises/concept/none/.docs/introduction.md:
```diff
@@ -0,0 +1,32 @@
+In Python, `None` is frequently used to represent the absence of a value, it can also be used to define a `null` variable or an object.
```
```suggestion
In Python, `None` is frequently used to represent the absence of a value --  a placeholder to define a `null` (empty) variable, object, or argument.
```

BethanyG commented on languages/python/exercises/concept/none/.docs/introduction.md:
```diff
@@ -0,0 +1,32 @@
+In Python, `None` is frequently used to represent the absence of a value, it can also be used to define a `null` variable or an object.
+
+If you have heard about `NULL` type in other languages then this must be familiar. `None` helps you to declare variables. These variables can be used later to be defined to a value. We can assign `None` to any variable.
```
```suggestion
If you've heard about or used the `NULL` or `nil` type in other languages,  then this usage of `None` in Python will be familiar to you. `None` helps you to declare variables that you don't yet have values for. These variables can then be re-assigned to a value later as needed. We can assign `None` to any variable, argument, or attribute.
```

BethanyG commented on languages/python/exercises/concept/none/.docs/introduction.md:
````diff
@@ -0,0 +1,32 @@
+#=> returns None
+```
+
+`None` evaluates to `False` when used in a conditional check.
````
```suggestion
`None` will evaluate to `False` when used in a conditional check, so it is useful for validating the "presence of" or "absence of" a value - _any_ value -- a pattern frequently used when a function or process might hand back an error object or message.  
```

mohanrajanr commented:
@BethanyG : Did the changes you requested. Kindly take a look when you have time.

PS: There are a lot of files that needs linting with prettier. Everytime i run prettier i see a lot of files
```
        modified:   languages/go/exercises/concept/booleans/.docs/introduction.md
        modified:   languages/javascript/exercises/concept/array-analysis/babel.config.js
        modified:   languages/javascript/exercises/concept/arrays/babel.config.js
        modified:   languages/javascript/exercises/concept/arrays/example.js
        modified:   languages/javascript/exercises/concept/basics/babel.config.js
        modified:   languages/javascript/exercises/concept/booleans/babel.config.js
        modified:   languages/javascript/exercises/concept/booleans/booleans.js
        modified:   languages/javascript/exercises/concept/booleans/booleans.spec.js
        modified:   languages/javascript/exercises/concept/booleans/example.js
        modified:   languages/javascript/exercises/concept/closures/babel.config.js
        modified:   languages/javascript/exercises/concept/nullability/babel.config.js
        modified:   languages/javascript/exercises/concept/numbers/babel.config.js
        modified:   languages/javascript/exercises/concept/numbers/example.js
        modified:   languages/javascript/exercises/concept/numbers/numbers.spec.js
        modified:   languages/javascript/exercises/concept/promises/babel.config.js
        modified:   languages/javascript/exercises/concept/promises/example.2.js
        modified:   languages/javascript/exercises/concept/promises/example.js
        modified:   languages/javascript/exercises/concept/promises/global.d.ts
        modified:   languages/javascript/exercises/concept/promises/promises.js
        modified:   languages/javascript/exercises/concept/promises/promises.spec.js
        modified:   languages/javascript/exercises/concept/recursion/babel.config.js
        modified:   languages/javascript/exercises/concept/strings/babel.config.js
        modified:   languages/javascript/exercises/concept/strings/strings.spec.js
        modified:   languages/javascript/exercises/practice/string-iterables/example.js
        modified:   languages/javascript/scripts/scaffold/template/babel.config.js
        modified:   languages/javascript/scripts/scaffold/template/concept-name.js
        modified:   languages/javascript/scripts/scaffold/template/concept-name.spec.js
        modified:   languages/javascript/scripts/scaffold/template/example.js
```

Should I add them to the PR ( would not be wise i guess ) or raise it to an maintainer for the respective language?. Javascript seems to be widely off on it!.

SleeplessByte commented:
How are you running prettier? Running it as `prettier .` is probably what's causing this.

BethanyG commented:
Hi @mohanrajanr,

 As @SleeplessByte has commented above -- our guess is that you're running prettier on everything in your branch, rather than just the files you've changed (_the Python & markdown ones_).  And it is getting really excited about the `.js` config files.  And that means you probably have a local prettier configuration for `.js` that is different from the one exercism is using.

But I see it is also picking up markdown changes from `go`.  So I suspect that however you are refreshing your branch from upstream is triggering prettier to see those refreshes as "changes" it has to go back over.  But since those changes were already merged into `master`, there shouldn't be a need for any sort of re-formatting.

So there are some ways around this:

1.  Configure a hotkey for prettier to run on demand, and only use it on the files you change
2.   Configure prettier to run on save -- but only the specific files you've changed    
3.  Configure prettier to only look at files in the `language/python/` directory on save
4.  Use `/format` as a comment on the PR to trigger prettier to run as a github action.  (_I've had mixed success with this one_)

And when you are refreshing your branch from `upstream`, use `git rebase upstream/branch` to replay your changes over the ones you pull in from `master`.  That way, the changes you pull in are not flagged as "changed" in your branch, and hopefully that will help prettier to not get "excited" about "extra" files.

But you definitely should **not**  add those linting edits to this  PR -- those are changes from other languages you've picked up, and not changes you've made -- your file count should stay at 9.

BethanyG commented on languages/python/exercises/concept/none/.docs/introduction.md:
````diff
@@ -0,0 +1,32 @@
+#=> <class 'NoneType'>
+```
+
+Python will also return `None` from a function that doesn't already have a stated `return` value, making it easy to check if the function completed.
````
```suggestion
Python will also return `None` from a function that doesn't already have a stated `return` value, making it easy to check if the function completed without errors.
```

BethanyG replied:
Suggest a quick code example here like:

```python
def no_values_returned():
    pass

>>> print(no_values_returned())
None
```

BethanyG commented on languages/python/exercises/concept/none/.docs/introduction.md:
```diff
@@ -0,0 +1,32 @@
+In Python, `None` is frequently used to represent the absence of a value -- a placeholder to define a `null` (empty) variable, object, or argument.
+
+If you've heard about or used a `NULL` or `nil` type in another programming language, then this usage of `None` in Python will be familiar to you. `None` helps you to declare variables that you don't yet have values for. These variables can then be re-assigned to a value later as needed.
```
```suggestion
If you've heard about or used a `NULL` or `nil` type in another programming language, then this usage of `None` in Python will be familiar to you. `None` helps you to declare variables or function arguments that you don't yet have values for. These can then be re-assigned to specific values later as needed.
```

BethanyG commented on languages/python/exercises/concept/none/.docs/introduction.md:
````diff
@@ -0,0 +1,32 @@
+type(a)
+#=> <class 'NoneType'>
+```
+
````
So I think it might be goo to add an example here of `none` as a default value for an argument.  
Something like:

```python
def add_to_todos(new_task, todo_list=None):
  if todo_list is None:
    todo_list = []
    todo_list.append(new_task)
  return todo_list
```

BethanyG commented on languages/python/exercises/concept/none/.docs/after.md:
````diff
@@ -0,0 +1,28 @@
+
+a = None #=> Variable 'a' is reset to an empty state. the value of a is now absent.
+
+```
````
We might also want to add some discussion here around what you need to do when you can't use `None` as your only placeholder.  Check out this [None](https://realpython.com/null-in-python/) article from Real Python.

mohanrajanr replied:
I may need to come back to this one again

BethanyG commented on languages/python/exercises/concept/none/.docs/hints.md:
```diff
@@ -0,0 +1,27 @@
+## General
+
+- Don't worry about how the arguments are derived, just focus on combining the arguments to return the intended result.
```
```suggestion
In Python, `None` is frequently used as a placeholder to represent the **absence of a value** for a variable, object, or argument.
```

BethanyG commented on languages/python/exercises/concept/none/.docs/instructions.md:
```diff
@@ -0,0 +1,89 @@
+In this exercise, you will get the basic idea of how to use 'None'.
```
```suggestion

```

BethanyG commented on languages/python/exercises/concept/none/.meta/design.md:
```diff
@@ -0,0 +1,28 @@
+- create a `None` object via literal declaration
+- check if an object is `None`
+- use a `None` object as a "False statement"
+
```
```suggestion
- use `None` as a default parameter value
```

BethanyG commented on languages/python/exercises/concept/none/.docs/instructions.md:
```diff
@@ -0,0 +1,89 @@
+
+The exercise has 6 challenges to complete.
+
+You are the Maître D' of a hotel, Your tasks are given based on the number of people who are waiting to eat at your hotel. Complete the tasks accordingly and proceed.
```
```suggestion
You are the Maître D' of a hotel restaurant.   Your task is to manage the seating arrangements for the dining room according to the number of seats available today, number of reservations,  and the "walk in" guests currently waiting to be seated.   For the purposes of this exercise, seating is assigned by first available empty seat.

You have 6 different dining room organization challenges to complete.
```

BethanyG commented on languages/python/exercises/concept/none/.docs/hints.md:
```diff
@@ -0,0 +1,27 @@
+## General
+
+- Don't worry about how the arguments are derived, just focus on combining the arguments to return the intended result.
+
```
```suggestion
## 1.  Make New Seating Chart

Remember, you will need to create an empty dictionary first.  Next, fill the dictionary _keys_ with however many seats are available and set their _values_ to a **placeholder** to indicate that they are available but unassigned.
```

BethanyG commented on languages/python/exercises/concept/none/.docs/hints.md:
```diff
@@ -0,0 +1,27 @@
+
+- Don't worry about how the arguments are derived, just focus on combining the arguments to return the intended result.
+
+## 1. Arrange the Seats
```
```suggestion
## 2. Arrange Reservations
```

BethanyG commented on languages/python/exercises/concept/none/.docs/hints.md:
```diff
@@ -0,0 +1,27 @@
+
+## 1. Arrange the Seats
+
+- You need to create a dict object first. Then you can loop from 1 to the number of guests and assign them to None
```
```suggestion
- If there isn't a guest list, you will want to return a `new_seating_chart()` filled with placeholders.  A `default argument` for your function might be helpful here.   If you do have a guest list, you can start with a `new_seating_chart()`, then loop from 1 to the number of guests and assign them to seats in order.
```

BethanyG commented on languages/python/exercises/concept/none/.docs/hints.md:
```diff
@@ -0,0 +1,27 @@
+
+- You need to create a dict object first. Then you can loop from 1 to the number of guests and assign them to None
+
+## 2. Is the Seat Empty?
```
```suggestion

```

BethanyG commented on languages/python/exercises/concept/none/.docs/hints.md:
```diff
@@ -0,0 +1,27 @@
+
+## 2. Is the Seat Empty?
+
+- You can check a variable is None by the if statements. `if var == None` or `if var is None` will tell you if the variable is None.
```
```suggestion
- You can check that a variable or value is None through an `if statement`. `if var is None` will  return True if the value is `None`.
```

BethanyG commented on languages/python/exercises/concept/none/.docs/hints.md:
```diff
@@ -0,0 +1,27 @@
+
+- You can check a variable is None by the if statements. `if var == None` or `if var is None` will tell you if the variable is None.
+
+## 3. Find the Empty seat
```
```suggestion
## 4. Current seating capacity
```

BethanyG commented on languages/python/exercises/concept/none/.docs/hints.md:
```diff
@@ -0,0 +1,27 @@
+
+## 3. Find the Empty seat
+
+- You can use the is_seat_empty() function through a loop of all the key, values in the dict object. When the value is none, you need to return the key.
```
```suggestion
- You can loop through all of the values in the dict object by calling `dict.values()`. Seats are available when their value is `None`.
```

BethanyG commented on languages/python/exercises/concept/none/.docs/hints.md:
```diff
@@ -0,0 +1,27 @@
+
+- You can use the is_seat_empty() function through a loop of all the key, values in the dict object. When the value is none, you need to return the key.
+
+## 4. Current Empty Seating Capacity
+
+- You need to find the number of empty seats ( that is pointing to `None`) in the dict object. Once you get the count, that is the answer
+
```
```suggestion

```

BethanyG commented on languages/python/exercises/concept/none/.docs/hints.md:
```diff
@@ -0,0 +1,27 @@
+
+- You need to find the number of empty seats ( that is pointing to `None`) in the dict object. Once you get the count, that is the answer
+
+## 5. Should we wait?
+
+- You need to find the current number of empty seats and find if it is greater than or equal to the number of guests waiting.
+
```
```suggestion
## 5. Accommodate Waiting Guests

- You need to find the current number of empty seats and check to see if it is greater than or equal to the number of guests waiting.  If the guests can be accommodated, you will need to call `find_all_available_seats()` to get a list of seat numbers you can assign guests to.  Remember that `range()` can take any sort of number as an argument....**including** the number returned from calling `len()` on a list.  Also remember that using `list[index_number]` will return the **value** that is located at the **index number** inside the brackets.

```

BethanyG commented on languages/python/exercises/concept/none/.docs/hints.md:
```diff
@@ -0,0 +1,27 @@
+
+- You need to find the current number of empty seats and find if it is greater than or equal to the number of guests waiting.
+
+## 6. Empty a seat
+
+- Given the seat number ( the key to the `dict` object ) and the seats dict object, you need to assign that particular seat (key) to value `None`.
```
```suggestion
## 6. Empty the Seats

- Given the seating chart `dict`, and a `list` of seat numbers, you'll want to indicate the seat is now available for another guest by replacing their name with a placeholder.   Looping through the seat number list and looking for those seats in the dictionary might be helpful here.
```

BethanyG commented on languages/python/exercises/concept/none/.docs/instructions.md:
```diff
@@ -0,0 +1,89 @@
+
+You are the Maître D' of a hotel, Your tasks are given based on the number of people who are waiting to eat at your hotel. Complete the tasks accordingly and proceed.
+
+## 1. Arrange the Seats
+
+Define the `arrange_the_seats()` function with 1 parameter. You know the number of guests who have reserved a table today.
+
+The task of this function is to return a `dict` object containing the seat number (in `int`) as the key and `None` as the value for the number of guests who are going to attend.
+
```
```suggestion
## 1.  Make Today's Seating Chart

Define the `new_seating_chart()` function that takes a size argument representing the number of seats that will be set in the dining room today.  If no `size` is given, the function should return a seating chart with 22 seats.  Seat values should have a placeholder of `None` to indicate they are available to assign to a guest.

## 2. Arrange Reservations

Define the `arrange_reservations()` function with 1 parameter for a list of guest names. This represents the number of people who've reserved places in the dining room today.

This function should return a `dict` seating chart of default size (22 seats), with guests assigned to seats in the order they appear on the reservation list.  All unassigned seats should be set to `None`.  If there are no guests, an "empty" seating chart with all `None` placeholders should be returned.

```

BethanyG commented on languages/python/exercises/concept/none/.docs/instructions.md:
````diff
@@ -0,0 +1,89 @@
+
+The task of this function is to return a `dict` object containing the seat number (in `int`) as the key and `None` as the value for the number of guests who are going to attend.
+
+```python
+arrange_the_seats(guests=4)
+#=> {1: None, 2: None, 3: None, 4: None}
+```
````
```suggestion
```python
arrange_reservatons(guests=["Walter", "Frank", "Jenny", "Carol", "Alice", "George"])
#=> {1: 'Walter', 2: 'Frank', 3: 'Jenny', 4: 'Carol', 5: 'Alice', 6: 'George', 7: None, 8: None, 9: None, 10: None, 11: None, 12: None, 13: None, 14: None, 15: None, 16: None, 17: None, 18: None, 19: None, 20: None, 21: None, 22: None}```
```

BethanyG commented on languages/python/exercises/concept/none/.docs/instructions.md:
````diff
@@ -0,0 +1,89 @@
+#=> {1: None, 2: None, 3: None, 4: None}
+```
+
+## 2. Is the Seat Empty?
+
+Define the `is_seat_empty()` function that takes 1 parameter. First parameter will contain the value of the seat which is checked before assigning to a guest.
+
+If the seat is empty, It will be of `None` value, else it will be mentioned with string `Occupied`.
+
+If the seat is empty, return `True` else return `False`.
+
+```python
+is_seat_empty(seat_value="Occupied")
+# => False
+is_seat_empty(seat_value=None)
+# => True
+```
+
````
```suggestion
## 3.  Find All the "Empty" Seats

Define the `find_all_available_seats(seats)` function that takes 1 parameter (_a seating chart dictionary_) and returns a `list` of seat numbers that are available for guests that are currently waiting.

If a seat is empty, It will be of `None` value in the dictionary.   Occupied seats will have the name of the guest.

```python
find_all_available_seats(seats)
>>>[1,16,5,7,3]
```

BethanyG commented on languages/python/exercises/concept/none/.docs/instructions.md:
````diff
@@ -0,0 +1,89 @@
+# => True
+```
+
+## 3. Find the Empty seat
+
+Define the `find_the_seat()` function that takes 1 parameter. First parameter will contain the `dict` object that is created by you while arranging the seats.
+
+Find the first empty seat that you come across in the list of seats passed to you.
+
+Your job is to return the seat number ( the key of the `dict` object ) that is passed to you.
+
+Note: There can be more than one seat which can be empty in the seats sent to you as a parameter.
+
+Bonus Tip: You can use the Function `is_seat_empty()` as an helper for this function.
+
+```python
+find_the_seat(seats={1: "Occupied", 2: None, 3: "Occupied"})
+# => 2
+```
````
```suggestion

```

BethanyG commented on languages/python/exercises/concept/none/.docs/instructions.md:
````diff
@@ -0,0 +1,89 @@
+# => 2
+```
+
+## 5. Should we wait?
+
+Define the `can_accomodate_seats()` function that takes two parameters. The first parameter will list all the seats. The second parameter is the seats you need to fill the guests who have come unexpected.
+
+You need to find out how many seats are empty, Based on the number of seats you need to find out whether you can give the unannounced guests the seats they want.
+
+You need to return True, if you have the sufficient number of seats, else False.
+
+Tip: You can use the curr_empty_seat_capacity() function as your helper.
+
+```python
+can_accomodate_seats(seats={1: "Occupied", 2: None, 3: "Occupied"}, guests=2)
+# => False
+can_accomodate_seats(seats={1: "Occupied", 2: None, 3: None}, guests=2)
+# => True
+```
+
````
```suggestion
## 5. Should we wait?

Define the `accommodate_waiting_guests(seats, guests)` function that takes two parameters. The first parameter will be a seating chart `dict`. The second parameter will be a `list` of guests who have "walked in" unexpectedly.  

You'll first need to find out how many seats are available and whether or not you can even give the unannounced guests seats at this time.  If you do not have enough seats, return `False`.

If seats are available, you will want to give the guests places on the seating chart, and return it updated.

Tip: You can use previously defined functions to do the calculations for you.

```python
starting_reservations = {1: 'Carol', 2: 'Alice', 3: 'George', 4: None, 5: None, 6: None, 7: 'Frank', 8: 'Walter'}
accomodate_guests(starting_reservations, ["Mort", "Suze", "Phillip", "Tony"]))
>>>False

starting_reservations = {1: None, 2: None, 3: None, 4: 'Carol', 5: 'Alice', 6: 'George', 7: None, 8: None, 9: None, 10: None, 11: None, 12: None, 13: None, 14: None, 15: None, 16: None, 17: None, 18: 'Frank', 19:  'Jenny', 20: None, 21: None, 22: 'Walter'}
accomodate_guests(starting_reservations, ["Mort", "Suze", "Phillip", "Tony"]))
>>>{1: 'Mort', 2: 'Suze', 3: 'Phillip', 4: 'Carol', 5: 'Alice', 6: 'George', 7: 'Tony', 8: None, 9: None, 10: None, 11: None, 12: None, 13: None, 14: None, 15: None, 16: None, 17: None, 18: 'Frank', 19: 'Jenny', 20: None, 21: None, 22: 'Walter'}
```

```

BethanyG commented on languages/python/exercises/concept/none/.docs/instructions.md:
````diff
@@ -0,0 +1,89 @@
+# => True
+```
+
+## 6. Empty a seat
+
+Define the `empty_the_seat()` function that takes two parameters. The first parameter will list all the seats. The second parameter is the seat you need to empty, that is you need to assign the seat with None.
+
+Return the `dict` of seats sent as the parameter after updating the empty seat.
+
+```python
+empty_the_seat(seats={1: "Occupied", 2: None, 3: "Occupied"}, 3)
+# => {1: "Occupied", 2: None, 3: None}
+```
````
```suggestion
## 6. Empty Seats

Define the `empty_seats()` function that takes two parameters. The first parameter will be a seating chart dictionary. The second parameter is a list of seat numbers you need to "free up" or empty -- that is,  you need to assign the seat number value to `None`.

Return the `dict` of seats sent as the parameter after updating the empty seat values.

```python
empty_seats(seats={1: "Alice", 2: None, 3: "Bob", 4: "George", 5: "Gloria"}, seat_numbers=[5,3,1])
# => {1: None, 2: None, 3: None, 4: "George", 5: None}
```
```

BethanyG commented on languages/python/exercises/concept/none/.meta/design.md:
```diff
@@ -0,0 +1,28 @@
+
+- **none:**
+
+## Prerequisites
+
+- `builtin-types`
+
```
```suggestion
## Prerequisites

- `bracket notation`, 
- `dicts`, 
- `dict-methods`,
- `indexing`,
- `len()`, 
- `lists`, 
- `list-methods`,
- `loops(),
- `range()`
```

BethanyG commented on languages/python/exercises/concept/none/.meta/design.md:
```diff
@@ -0,0 +1,28 @@
+- [the null object](https://docs.python.org/3/library/stdtypes.html#the-null-object)
+- [edspresso_none](https://www.educative.io/edpresso/what-is-the-none-keyword-in-python)
+- [none_keyword](https://docs.python.org/3/library/constants.html#None)
+- [none_stackoverflow](https://stackoverflow.com/questions/19473185/what-is-a-none-value)
```
```suggestion
- [None (Stackoverflow)](https://stackoverflow.com/questions/19473185/what-is-a-none-value)
- [Null in Python:  Understanding None (Real Python)](https://realpython.com/null-in-python/)
```

BethanyG commented on languages/python/exercises/concept/none/.meta/example.py:
```diff
@@ -0,0 +1,33 @@
+def arrange_the_seats(guests):
+    seats = dict()
+    for seat_num in range(1, guests+1):
+        seats[seat_num] = None
+    return seats
+
```
```suggestion
def new_seating_chart(size=22):
    seats = dict()
    for number in range(1, size+1):
        seats[number] = None
    return seats

def arrange_reservations(guests=None):
    if guests is None:
        return new_seating_chart()
    else:
        seats = new_seating_chart()
        for seat_number in range(1, len(guests)):
            seats[seat_number] = guests[seat_number]
    return seats
```

BethanyG commented on languages/python/exercises/concept/none/.meta/example.py:
```diff
@@ -0,0 +1,33 @@
+        seats[seat_num] = None
+    return seats
+
+def is_seat_empty(seat_value):
+    if seat_value == None:
+        return True
+    else:
+        return False
+
+def find_the_seat(seats):    
+    for seat_num, value in seats.items():
+        if is_seat_empty(value):
+            return seat_num
```
```suggestion
def find_all_available_seats(seats):
    available = []

    for seat_num, value in seats.items():
        if value is None:
            available.append(seat_num)
    return available
```

BethanyG commented on languages/python/exercises/concept/none/.meta/example.py:
```diff
@@ -0,0 +1,33 @@
+
+def curr_empty_seat_capacity(seats):
+    count = 0
+    for seat_num, value in seats.items():
+        if is_seat_empty(value):
+            count = count + 1
+    return count
```
```suggestion
    for value in seats.values():
        if value is None:
            count += 1
    return count
```

BethanyG commented on languages/python/exercises/concept/none/.meta/example.py:
```diff
@@ -0,0 +1,33 @@
+            count = count + 1
+    return count
+
+def can_accomodate_seats(seats, guests):
+    curr_empty_seats = curr_empty_seat_capacity()
+    if guests > curr_empty_seat_capacity:
+        return False
+    else:
+        return True
```
```suggestion
def accommodate_waiting_guests(seats, guests):
    curr_empty_seats = current_seating_capacity(seats)
    empty_seat_list = find_all_empty_seats(seats)

    if len(guests) > curr_empty_seats:
        return False
    else:
        for index in range(len(guests)):
            seats[empty_seat_list[index]] = guests[index]
    return seats
```

BethanyG commented on languages/python/exercises/concept/none/.meta/example.py:
```diff
@@ -0,0 +1,33 @@
+    else:
+        return True
+
+def empty_the_seat(seats, seat_number):
+    seats[seat_number] = None
```
```suggestion
def empty_seats(seats, seat_numbers):
     for seat in seat_numbers:
        seats[seat_number] = None
```

BethanyG commented on languages/python/exercises/concept/none/none.py:
```diff
@@ -0,0 +1,17 @@
+def arrange_the_seats(guests):
+    pass
+
+def is_seat_empty(seat_value):
+    pass
+
+def find_the_seat(seats):
+    pass
+
+def curr_empty_seat_capacity(seats):
+    pass
+
+def can_accomodate_seats(seats, guests):
+    pass
+
+def empty_the_seat(seats, seat_number):
+    pass
```
```suggestion
def new_seating_chart():
    pass

def arrange_reservations():
    pass

def find_all_available_seats():
   pass

def current_seating_capacity():
   pass

def accommodate_waiting_guests():
   pass

def empty_seats():
    pass
```

BethanyG commented on languages/python/exercises/concept/none/none_test.py:
```diff
@@ -0,0 +1,85 @@
+        self.assertFalse(
+            can_accomodate_seats({1: "Occupied", 2: None, 3: "Occupied"}, 1),
+            msg="Cannot Accomodate Seats for the Input"
+        )
```
ok.  For this test file, I think we can file an "improvement issue", rather than dragging this out even more with more edits and feedback.  We just need to bring these up to date with the changed `example.py`

mohanrajanr replied:
I did modify the tests with the new changed functions though. We can still have an improvement exercise.!

BethanyG reviewed (approved):
LGTM!  Thank you for all your hard work on this - I know it was a lot!
